### PR TITLE
[net processing] Various tidying up of PeerManagerImpl ctor

### PR DESCRIPTION
**Pull request by jnewbery**
- Use default initialization of PeerManagerImpl members where possible
- Remove unique_ptr indirection where it's not needed

### CODE REVIEW

**Comment by fanquake** [comment]
Cirrus failures:
```bash
In file included from init.cpp:55:
./txorphanage.h:50:28: error: reading variable 'm_orphans' requires holding mutex 'g_cs_orphans' [-Werror,-Wthread-safety-analysis]
    size_t Size() { return m_orphans.size(); };
                           ^
1 error generated.
make[2]: *** [libbitcoin_server_a-init.o] Error 1
make[2]: *** Waiting for unfinished jobs....
  CXX      libbitcoin_server_a-net_processing.o
In file included from net_processing.cpp:29:
./txorphanage.h:50:28: error: reading variable 'm_orphans' requires holding mutex 'g_cs_orphans' [-Werror,-Wthread-safety-analysis]
    size_t Size() { return m_orphans.size(); };
                           ^
net_processing.cpp:410:59: error: expected ';' at end of declaration list
    CRollingBloomFilter m_recent_rejects{120000, 0.000001} GUARDED_BY(cs_main);
                                                          ^
                                                          ;
net_processing.cpp:429:73: error: expected ';' at end of declaration list
    CRollingBloomFilter m_recent_confirmed_transactions{48000, 0.000001} GUARDED_BY(m_recent_confirmed_transactions_mutex);
                                                                        ^
                                                                        ;
3 errors generated.
```

**Comment by practicalswift** [comment]
Strong Concept ACK

I find the suggested code easier to reason about from a safety and correctness perspective.

**Comment by MarcoFalke** [comment]
The build failure is because `Assume` breaks static lock analysis. One way to fix is to assign the assumed-value to a bool first.

**Comment by jnewbery** [comment]
Thanks Marco! I've just reverted these checks to `assert()` for now.

**Comment by hebasto** [comment]
@jnewbery Help me to understand the motivation for "[net processing] Move consistency checks to PeerManagerImpl dtor" commit please.

My current understanding is following. A consistency is an invariant that the `PeerManagerImpl` class must hold. It's ctor's job to establish class's invariants, and member functions' job to preserve invariants. Delegating such a task to dtor seems unusual for me.

**Comment by jnewbery** [comment]
> Help me to understand the motivation for "[net processing] Move consistency checks to PeerManagerImpl dtor" commit please.
> 
> My current understanding is following. A consistency is an invariant that the PeerManagerImpl class must hold. It's ctor's job to establish class's invariants, and member functions' job to preserve invariants. Delegating such a task to dtor seems unusual for me.

I think that's fair. My reasoning was that these invariant checks are all test/development asserts. We'd hope that they'd never be hit in the wild. They currently run when the peer count drops to zero, but if there's an inconsistency then, we'd expect those invariants to also be false when the PeerManagerImpl is destructed on shutdown.

My motivation is that I want to pull some of these members out from being guarded by cs_main. The global counts like `m_outbound_peers_with_protect_from_disconnect` and `m_wtxid_relay_peers` don't actually need to be atomically updated with the corresponding`CNodeState` member variables `m_chain_sync.m_protect` and `m_wtxid_relay`. We could pull those things out of `cs_main` and just make them separate atomics and that'd be safe, but not obviously so. `FinalizeNode()` is called by the same thread that increments the counters, so there's no way that they can be non-zero whilst the final peer is being removed. However, if other threads _were_ able to increment the counters, these invariants would no longer be true. That's not the case if we move the invariant checks to the destructor. Inside the dtor, we know that no other thread can be mutating the members, so the counters _must_ be at zero.

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1256,6 +1245,20 @@ PeerManagerImpl::PeerManagerImpl(const CChainParams& chainparams, CConnman& conn
     scheduler.scheduleFromNow([&] { ReattemptInitialBroadcast(scheduler); }, delta);
 }
 
+PeerManagerImpl::~PeerManagerImpl()
+{
+    // Do consistency checks on shutdown
+    LOCK(cs_main);
```
Comment (hebasto) [inline src/net_processing.cpp:1251]: Only one thread has an access to the object while a dtor is called. Why lock?

**Comment by jnewbery** [comment]
@hebasto - there are various changes to net_processing logging happening in 21527, so I've removed the changes to the destructor here until those land. Let me know if you think the remaining changes to the constructor in this PR are worth keeping.

Also rebased on current master.

**Comment by 78051301012** [comment]
ok

**Comment by Crypt-iQ** [comment]
Code review ACK b26a5adceaf4e1dbf35aa595f7870fe059940959

I was confused by the asserts prior to this PR.  Also there are two typos in the commit messages "contetx" should be "context"?

**Comment by jnewbery** [comment]
Thanks @Crypt-iQ. I've corrected the typos in the commit logs.

> I was confused by the asserts prior to this PR.

These were to ensure that `recentRejects` wasn't null before dereferencing it. No need for that now that `m_recent_rejects` is not a pointer.

**Comment by Crypt-iQ** [comment]
Code review ACK d6cef4d8a6a70920e98f40530113733142025f6a

Only the commit messages were fixed from b26a5adceaf4e1dbf35aa595f7870fe059940959, so should be good.

**Comment by jnewbery** [comment]
Rebased on master to fix fuzz CI timeout.

**Comment by Crypt-iQ** [comment]
Code review ACK 93e881f

Commit message 9e6bf785 has a typo: rejectRejects should be recentRejects

**Comment by jnewbery** [comment]
Rebased.

> Commit message 9e6bf78 has a typo: rejectRejects should be recentRejects

Fixed. Thank you!

**Comment by MarcoFalke** [comment]
review ACK d5dd3fa1de2f96a7ddfd8b300829c68b93bd4607 🔠

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK d5dd3fa1de2f96a7ddfd8b300829c68b93bd4607 🔠
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhpNwwAjxKP8RmX2H5ReC3tV5MQ/WvFMI9x+WHwIiNcpkRQ8wk4kErsemjnoroT
gEZg/nq3IfgMQL3LvaU270gGOdH3J45TFJtoGCtq+nogV/nU8TwjlCzP0vUsRzfP
3hiarnsn3GIJs9r01uC31GII/TrTSZPDjsKxd9X5aLKAtjsMlB9C5mZdQtQ5CYul
+3QGhnGGZpTIqE9V24dY0sOUrO54UtSW87qaxRXx88MkgVz+LTF1TDjSJy9We6Hj
ehe8OTC9MoHOYJ5rv7aiLsrJ9H28Ntal6w84Q9lts1laTEyMlazegtjIvoUogrP4
eswz53RfJg1APyIDZPn62+xNp3XYIG7ib29L3A/lsjjbw6d862EDOmyjVVloTMcM
Tb7ruvFaZdr8k9TBeKtZ1P/ecEPszUWldCIJ5RXo9iIIGaks7f8OKJNyTmgxjfVj
HOVgn1gwuE2sTQHb6pSuW45JBOslXCrYaCOQNWrCpX29X0iOpT4bJM31WVX1lFzH
TtiamXSG
=v+4O
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `cb3d358ad6eccc7781819f9917bb2f9a6f7b802e636f39fe7a8d779c985d5dfc  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108cb3d358ad6eccc7781819f9917bb2f9a6f7b802e636f39fe7a8d779c985d5dfcf010316e3ddb097cd337ab10961b66d4eac608fff0100715596d8b87016bf50cbc9753e306b108f0204d99e4e491994a753b2e4d4d8e7da2a7d5704d4219e58b79ef961171377a9ba308f0204e03067616aa017970d12f609e73e27f83dcf2ff4bac312dc254338e2a98c49408f10460c7877df00848158bafbbded5320083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01010d3b070c6f5eebb7eadafd70d6ae03608f10460c7877ef00803657fd02cfa2cce0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010cc92ec31106dbae1657dcdc1776063a808f120cd1ce0e1901197af06c9d9624dcde444371326d155c41ee3a538d9aa22b6ab5108f10460c7877df0080219a75b42df73de0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010d47256d9cab9eee744ea062efb927c2208f10460c7877ef0086e50d6304a395c0d0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/txorphanage.h [hunk 2]
```diff
@@ -47,6 +47,8 @@ class TxOrphanage {
      * (ie orphans that may have found their final missing parent, and so should be reconsidered for the mempool) */
     void AddChildrenToWorkSet(const CTransaction& tx, std::set<uint256>& orphan_work_set) const EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
 
+    size_t Size() { return WITH_LOCK(g_cs_orphans, return m_orphans.size(););};
```
Comment (hebasto) [inline src/txorphanage.h:50]: 60cf71cfc16ec59857bd97608f599f8d2346c23c

1) style-nit: redundant `;` and missed space:
```suggestion
    size_t Size() { return WITH_LOCK(g_cs_orphans, return m_orphans.size()); };
```

2) what about `LOCKS_EXCLUDED(::g_cs_orphans)` thread-safety annotation?

3) style-nit: isn't `LOCK(::g_cs_orphans); return m_orphans.size();` shorter and more readable?

4) why not use global namespace `::` for global mutexes consistently?

File: src/net_processing.cpp [hunk 3]
```diff
@@ -379,7 +379,8 @@ class PeerManagerImpl final : public PeerManager
     /** The height of the best chain */
     std::atomic<int> m_best_height{-1};
 
-    int64_t m_stale_tip_check_time; //!< Next time to check for stale tip
+    /** Next time to check for stale tip */
+    int64_t m_stale_tip_check_time{0};
```
Comment (hebasto) [inline src/net_processing.cpp:383]: 754bdaa61fe6473ca7cf9e89146eb8419fdd7c0a

I'm not an expert in Doxygen comments, but our [Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-doxygen-compatible-comments) suggest `//!` or `//!<` for variables.

File: src/net_processing.cpp [hunk 4]
```diff
@@ -452,16 +453,26 @@ class PeerManagerImpl final : public PeerManager
      *
      * Memory used: 1.3 MB
      */
-    std::unique_ptr<CRollingBloomFilter> recentRejects GUARDED_BY(cs_main);
+    CRollingBloomFilter m_recent_rejects GUARDED_BY(cs_main) {120000, 0.000001};
```
Comment (hebasto) [inline src/net_processing.cpp:456]: 6e48bfe706934fa4b71a3021d979321e94343c21, style nit:
```suggestion
    CRollingBloomFilter m_recent_rejects GUARDED_BY(::cs_main){120'000, 0.000001};
```

([`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy) states that the space is redundant)

File: src/net_processing.cpp [hunk 5]
```diff
@@ -452,16 +453,26 @@ class PeerManagerImpl final : public PeerManager
...
+     * If the number of transactions appearing in a block goes up, or if we are
+     * seeing getdata requests more than an hour after initial announcement, we
+     * can increase this number.
+     * The false positive rate of 1/1M should come out to less than 1
+     * transaction per day that would be inadvertently ignored (which is the
+     * same probability that we have in the reject filter).
      */
     Mutex m_recent_confirmed_transactions_mutex;
-    std::unique_ptr<CRollingBloomFilter> m_recent_confirmed_transactions GUARDED_BY(m_recent_confirmed_transactions_mutex);
+    CRollingBloomFilter m_recent_confirmed_transactions GUARDED_BY(m_recent_confirmed_transactions_mutex) {48000, 0.000001};
```
Comment (hebasto) [inline src/net_processing.cpp:475]: d5dd3fa1de2f96a7ddfd8b300829c68b93bd4607, style-nit:
```suggestion
    CRollingBloomFilter m_recent_confirmed_transactions GUARDED_BY(m_recent_confirmed_transactions_mutex){48'000, 0.000001};
```

**Review by hebasto — APPROVED** [review]
ACK d5dd3fa1de2f96a7ddfd8b300829c68b93bd4607

While touching the `net_processing.cpp`, also a typo which was introduced in #22141 could be fixed:
https://github.com/bitcoin/bitcoin/blob/5c2e2afe990ad6e1cc8b08513fa2d853d497dace/src/net_processing.cpp#L470

`recieved` ==> `received`

**Comment by practicalswift** [comment]
cr ACK d5dd3fa1de2f96a7ddfd8b300829c68b93bd4607

File: src/net_processing.cpp (hunk 3 above)
Comment (jnewbery) [reply src/net_processing.cpp:383]: Hmm, I wasn't aware of that guidance in the developer notes. Perhaps it should be updated - `/** comment */` is perfectly acceptable to doxygen, and many of the variables/members in the code base use that style.

File: src/txorphanage.h (hunk 2 above)
Comment (jnewbery) [reply src/txorphanage.h:50]: All fixed. Thanks for the suggestions!

File: src/net_processing.cpp (hunk 4 above)
Comment (jnewbery) [reply src/net_processing.cpp:456]: Done. Thanks!

File: src/net_processing.cpp (hunk 5 above)
Comment (jnewbery) [reply src/net_processing.cpp:475]: Done

**Comment by jnewbery** [comment]
Thanks for the reviews @MarcoFalke, @hebasto, @practicalswift. I've addressed all of the review comments from @hebasto so this is now ready for re-review.

> While touching the net_processing.cpp, also a typo which was introduced in #22141 could be fixed:
	
I was too slow. It's been fixed in #22323.

**Review by theStack — APPROVED** [review]
Code-review ACK 214e91a31b483c19a32ca22962e117a3e2688f86

one stylistic/readability-improvement nit: could also use the digit separators (') for the floating point numeral for the CRollingBloomFilter (valid since C++14 according to https://en.cppreference.com/w/cpp/language/floating_literal):

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 25fbd33b5..772ea2ced 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -453,7 +453,7 @@ private:
      *
      * Memory used: 1.3 MB
      */
-    CRollingBloomFilter m_recent_rejects GUARDED_BY(::cs_main){120'000, 0.000001};
+    CRollingBloomFilter m_recent_rejects GUARDED_BY(::cs_main){120'000, 0.000'001};
     uint256 hashRecentRejectsChainTip GUARDED_BY(cs_main);

     /*
@@ -472,7 +472,7 @@ private:
      * same probability that we have in the reject filter).
      */
     Mutex m_recent_confirmed_transactions_mutex;
-    CRollingBloomFilter m_recent_confirmed_transactions GUARDED_BY(m_recent_confirmed_transactions_mutex){48'000, 0.000001};
+    CRollingBloomFilter m_recent_confirmed_transactions GUARDED_BY(m_recent_confirmed_transactions_mutex){48'000, 0.000'001};

     /** Have we requested this block from a peer */
     bool IsBlockRequested(const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```

**Comment by jnewbery** [comment]
Thanks for the review @theStack. I didn't realize that the digit separator was permitted after the decimal point. I've taken your suggestion and rebased on master.

**Review by theStack — APPROVED** [review]
re-ACK fde1bf4f6136638e84cdf9806eedaae08e841bbf

checked via `git range-diff 214e91a3...fde1bf4f` that since my previous ACK, the only non-rebase-related change is using the digit separator also for floating points now, as suggested.

**Comment by MarcoFalke** [comment]
ACK fde1bf4f6136638e84cdf9806eedaae08e841bbf 👞

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK fde1bf4f6136638e84cdf9806eedaae08e841bbf 👞
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgIaAv+PZATwaBhcsf9/tfSEfMxRLP+BpK9c5vKV+URqEVG8wL+rV+vi/4Sdv5K
C7aNNt2AXAYZgqnIAvmsg1uL8G3uFZHHAo54ziHVjS7MlrG0+SSg5jKOerW/QjwA
Qb1BJqMP/ra4Jm7PHVcbGTjQg6Z0t3kJ+11jzqpVCRJ21NTPNwajBrrcpNKkdwHP
7sIKPVi5LLvK8V43gwGoAbEl9Dtwj/SfOIkaINpeZv8URYeKLYey3s3yjiO06wH1
9eXj+Sglww9GiPETOBuFMEMKcc9hCktK2oEHMv3U3d5cC12F9Dtjp1NazPNmBnvj
FwNAVbP//HwTIv0oZ+7dsZ/SKNIgKlgXR8fAsKH0ZiD85vC1ZalobhVqkmrRMX86
vW9ZP2Fj8X36Ez6vD+xRvy0cFMs+oQPPE4VLBnysmB5qT8uzcLSvSLTdwklW7FWK
T8XSw4GuNv/l/R7lEARBrkzajpMxtxRxNzvI5Zb7mu5CQWTZxogQFbCji+Vnk7ok
R+vr6Dyk
=9Rfq
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `d2de22e790c8247f5ed87c9b394b8c69ce8ccefa133cb79fea5a41603e328f92  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108d2de22e790c8247f5ed87c9b394b8c69ce8ccefa133cb79fea5a41603e328f92f010e3ac760781a3039b3f1e23341a4490c008fff01011934f625394ceb8051d36282fd0093f08f104610169f1f008025bce0f3c7a8c000083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010655d5806e9368d83332a04c7e5505ed308f104610169f1f0089f0d0440f0ef060b0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a3413e59c399c36333b293a034b4b2af08f104610169f1f00841c7b79c31036c130083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e84f3d4b09cf6ce332a037fd500eb28d08f104610169f1f008998ee3bdbb423df80083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

**Pull request merged**